### PR TITLE
[NHC] Use a single Evaluator trait

### PR DESCRIPTION
**Pull request by banool**
## Description
Previously the way I was differentiating between different evaluator types was by using different traits. These had `evaluate` methods that took in different arguments, returned different errors, etc. This led to a lot of code duplication, both for defining evaluators as well as building and calling them. In this PR I switch over to using a single Evaluator trait where the input and errors are associated types. To facilitate this I introduce an EvaluatorType enum that stores boxed trait objects to know which inputs to pass to which evaluator implementations. See the code comments for more info.

## Test Plan
See test plan from https://github.com/aptos-labs/aptos-core/pull/1454.

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/1542)

### CODE REVIEW

**Comment by banool** [comment]
These files are the real bread and butter of the PR:
- `ecosystem/node-checker/src/evaluator/traits.rs`
- `ecosystem/node-checker/src/evaluator/traits.rs`
- `ecosystem/node-checker/src/evaluators/build_evaluators.rs`
- `ecosystem/node-checker/src/runner/blocking_runner.rs`

**Review by gregnazario — COMMENTED** [review]
I didn't see anything major, but there's a lot going on here, I'd have to take a second pass on it when I have more time

File: ecosystem/node-checker/src/configuration/create.rs [hunk 1]
```diff
@@ -24,6 +22,9 @@ pub struct Create {
 
     #[clap(short, long, arg_enum, default_value = "yaml")]
```
[the remark this answers is not in the document]
Comment (banool) [reply ecosystem/node-checker/src/configuration/create.rs:23]: Sure.

File: ecosystem/node-checker/src/configuration/types.rs [hunk 2]
```diff
@@ -99,13 +101,13 @@ impl NodeConfiguration {
...
-    pub build_version_evaluator_args: BuildVersionEvaluatorArgs,
+    pub consensus_proposals_args: ConsensusProposalsEvaluatorArgs,
 
     #[clap(flatten)]
-    pub consensus_evaluator_args: ConsensusMetricsEvaluatorArgs,
+    pub state_sync_version_args: StateSyncVersionEvaluatorArgs,
 
     #[clap(flatten)]
-    pub state_sync_evaluator_args: StateSyncMetricsEvaluatorArgs,
+    pub system_information_build_version_args: SystemInformationBuildVersionEvaluatorArgs,
```
Comment (JoshLind) [inline ecosystem/node-checker/src/configuration/types.rs:110]: nit: this is a super long name 😨

File: ecosystem/node-checker/src/evaluator/traits.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    /// We require this function to ensure we can build all evaluators
+    /// from the top level evaluator args.
+    fn from_evaluator_args(evaluator_args: &EvaluatorArgs) -> Self
+    where
+        Self: Sized;
+}
+
+/*
+/// todo: explain that this function is not on the trait because it makes the
```
Comment (JoshLind) [inline ecosystem/node-checker/src/evaluator/traits.rs:55]: Which function are we talking about here?

File: ecosystem/node-checker/src/evaluator/traits.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+// not have functions or methods that do not take self as a parameter unless
+// they are marked as `where Self: Sized`. This constraint means these functions
+// can not be called on the trait object itself, just on the struct / Self.
+// For more information on this topic, see:
+// https://doc.rust-lang.org/reference/items/traits.html#object-safety
+
+/// An Evaluator is a component of NHC that is responsible for evaluating
+/// a particular aspect of the node under investigation, be that metrics,
+/// system information, API checks, load tests, etc.
+#[async_trait::async_trait]
```
Comment (JoshLind) [inline ecosystem/node-checker/src/evaluator/traits.rs:21]: Silly question: where did the async requirement come from?

**Review by JoshLind — APPROVED** [review]
Definitely a lot less boilerplate now. 😄 Also, I think some of the comments could be compacted (a few things are obvious and the text is slightly verbose at times).

File: ecosystem/node-checker/src/evaluator/traits.rs (hunk 4 above)
Comment (banool) [reply ecosystem/node-checker/src/evaluator/traits.rs:21]: For future evaluators where they need to do something async, e.g. the TPS evaluator, latency evaluator, etc. Up until now the evaluators had what they needed passed in, but other evaluators don't necessarily work like that.

File: ecosystem/node-checker/src/configuration/types.rs (hunk 2 above)
Comment (banool) [reply ecosystem/node-checker/src/configuration/types.rs:110]: Agreed, let me trim that.

**Pull request merged**